Topics/Development_Process/Error_Handling_in_ExpressJS.md

```diff
@@ -0,0 +1,184 @@
+```
+/** @const  {Object}  [STATUS_CODES] Maps status labels to status codes. */
+const  STATUS_CODES = {
+	SUCCESS:  200,
+	CREATED:  201,
+	INVALID:  400,
+	UNAUTHORISED:  401,
+	FORBIDDEN:  403,
+	NOT_FOUND:  404,
+	CONFLICT:  409,
+	PAYLOAD_TOO_LARGE:  413,
+	UNSUPPORTED_MEDIA_TYPE:  415,
+	TEAPOT:  418,
+	FAILED:  500,
+	NOT_IMPLEMENTED:  501,
+	INSUFFICIENT_STORAGE:  507
+}
+```
+### Implementing a Custom Error class
+
+Implementing a custom error class is as simple as extending `Error`, calling the parent constructor via `super(message)` in the `CustomError` constructor and then adding any additional attributes and methods as needed. In the below example, we have added a status code attribute which can be passed into the constructor after the error message to associate that error with an appropriate HTTP status code.
+
+***customErrors.js***
+```
+const STATUS_CODES = { ... } // See above subsection
+
+class CustomError extends Error {
+
+    /**
+     * Create a Custom Error.
+     * @param {string} [message] A descriptive message for the error.
+     * @param {number} [status_code = DEFAULT_STATUS_CODE] The corresponding status code for this error.
+     */
+    constructor(message, status_code = DEFAULT_STATUS_CODE) {
+        super(message);
+        this.name = this.constructor.name;
+        Error.captureStackTrace(this, this.constructor);
+        this._status_code = status_code;
+    }
+
+    /**
+     * Get the status code for this error.
+     * @return {number} The status code for this error.
+     */
+    get status_code() {
+        return this._status_code;
+    }
+
+    /**
+     * Set the status code for this error.
+     * @param {number} [new_status_code] The new status code for this error.
+     */
+    set status_code(new_status_code) {
+        this._status_code = new_status_code;
+    }
+}
+
+module.exports = {STATUS_CODES, CustomError};
+```
+
+### Example Usage
+Instead of throwing a generic `Error` object in our API logic, we can now throw our `CustomError` object. In the example below we show how we can throw an informative error for a attempting to get a resource that does not exist.
+
+***routes.js***
+```
+const db = require('./dbPlugin'); // A plugin for DB requests
+const {CustomError, STATUS_CODES} = require('./customErrors');
+
+const routes = (app) => {
+
+    app.get('/resource/:resource_id/', async (req, res) => {
+
+        const resource_id = req.params.resource_id;
+        
+        try {
+            const my_resource = await db.getResource(resource_id);
+
+            if (my_resource == null) {
+
+                // This indicates the requested resource was not found in the DB
+
+                throw new CustomError(
+                    `Resource with id=${resource_id} not found`,
+                    STATUS_CODES.NOT_FOUND // i.e. 404
+                );
+            }
+            
+            res.status(STATUS.SUCCESS); // Set success status code 200
```

Comment:
I think this is a typo. It should be STATUS_CODES.SUCCES if I'm not mistaken.

Reply:
Good spot!